File: agents/issue-reconciler/spokes/openspec-evidence.ts
import type { GitHubClient } from '../api/client.ts';
import { fetchOpenSpecChanges } from '../api/fetchers/repo.ts';

const ISSUE_REF_RE = /#(\d+)/g;

/**
 * Correlates each in-flight OpenSpec proposal to the issue number(s) it
 * references (e.g. "Closes #12"), returning issue number -> change names.
 * One repo-wide fetch, not one per issue - the proposal list is small and
 * shared across the whole batch.
 */
export async function gatherOpenSpecProposals(client: GitHubClient): Promise<Map<number, string[]>> {
  const changes = await fetchOpenSpecChanges(client);
  const byIssue = new Map<number, string[]>();

  for (const change of changes) {
    const referenced = new Set<number>();
    for (const match of change.proposalText.matchAll(ISSUE_REF_RE)) {
      referenced.add(Number(match[1]));
    }

    for (const issueNumber of referenced) {
      const existing = byIssue.get(issueNumber) ?? [];
      existing.push(change.name);
      byIssue.set(issueNumber, existing);
    }
  }

  return byIssue;
}

File: agents/issue-reconciler/issue_reconciler/spokes/openspec_evidence.py
from __future__ import annotations

import re

from issue_reconciler.api.client import GitHubClient
from issue_reconciler.api.fetchers.repo import fetch_openspec_changes

_ISSUE_REF_RE = re.compile(r"#(\d+)")


def gather_openspec_proposals(client: GitHubClient) -> dict[int, list[str]]:
    """Correlates each in-flight OpenSpec proposal to the issue number(s) it
    references (e.g. "Closes #12"), returning issue number -> change names.
    One repo-wide fetch, not one per issue - the proposal list is small and
    shared across the whole batch.
    """
    changes = fetch_openspec_changes(client)
    by_issue: dict[int, list[str]] = {}

    for change in changes:
        referenced = {int(n) for n in _ISSUE_REF_RE.findall(change["proposal_text"])}
        for issue_number in referenced:
            by_issue.setdefault(issue_number, []).append(change["name"])

    return by_issue
